src/main/java/com/google/summit/ast/declaration/FieldDeclaration.kt

```diff
@@ -18,24 +18,47 @@ package com.google.summit.ast.declaration
 
 import com.google.summit.ast.Identifier
 import com.google.summit.ast.Node
+import com.google.summit.ast.NodeWithSourceLocation
 import com.google.summit.ast.SourceLocation
 import com.google.summit.ast.TypeRef
 import com.google.summit.ast.expression.Expression
+import com.google.summit.ast.modifier.Modifiable
+import com.google.summit.ast.modifier.Modifier
+
+/**
+ * A group of comma-separated [FieldDeclaration]s that share the same [type] and [modifiers].
+ *
+ * @property type a reference to the type of the field
+ * @param loc the location in the source file
+ */
+class FieldDeclarationGroup(val type: TypeRef, loc: SourceLocation) :
+  NodeWithSourceLocation(loc), Modifiable {
+
+  /** The fields declared in this group. */
+  lateinit var declarations: List<FieldDeclaration>
+
+  override var modifiers: List<Modifier> = emptyList()
+
+  override fun getChildren(): List<Node> = modifiers + type + declarations
+}
 
 /**
  * A declaration for a field (a.k.a. class member).
  *
  * @param id the name of the field
- * @property type a reference to the type of the field
  * @property initializer an optional initializer expression
- * @param sourceLocation the location in the source file
+ * @param group the [FieldDeclarationGroup] that contains this field
+ * @param loc the location in the source file
  */
 class FieldDeclaration(
   id: Identifier,
-  val type: TypeRef,
   val initializer: Expression?,
+  group: FieldDeclarationGroup,
   loc: SourceLocation
-) : Declaration(id, loc) {
+) : Declaration(id, loc), Modifiable by group {
```

Comment:
`FieldDeclaration` no longer contains `List<Modifier>`, but it would be convenient for it to maintain its modifier interface instead of clients having to indirectly access modifiers through the parent `FieldDeclarationGroup`. `FieldDeclaration` would need to delegate to its `Modifiable` parent for this.

Here are the options I have thought of for this delegating logic:

---

1. Kotlin delegation - `Modifiable by group` (current)

Pros: concise; interface changes are automatically reflected
Cons:
- The `FieldDeclarationGroup` node has to be constructed before the `FieldDeclaration` children, which is unusual for `Translate`. `FieldDeclarationGroup.declarations` has to be `lateinit` and `var`.
- This may work for fields because both `FieldDeclaration` and `FieldDeclarationGroup` are currently constructed in [one translation method](https://github.com/google/summit-ast/pull/16/files#diff-b8ed44ff3e9cfc581f5dd3a149280a9c6c8d6fbede61c61363e93e3bb0069c36R443), but it may not work for other nodes. Nodes are typically constructed in separate translation methods and ANTLR seems to restrict what data can be passed around - it may not be possible to pass the parent node to children in other instances.
- `modifiers` is `var`, so it can be modified indirectly.

---

2. Mixin-style interface

e.g.
```kotlin
interface DelegatesModifiers {
  val modifierDelegate: Modifiable

  val modifiers
    get() = modifierDelegate.modifiers

  val annotationModifiers
    get() = modifierDelegate.annotationModifiers

  val keywordModifiers
    get() = modifierDelegate.keywordModifiers

  val hasKeyword
    get() = modifierDelegate::hasKeyword
}
```
where `FieldDeclaration` is
```kotlin
class FieldDeclaration : Declaration(), DelegatesModifiers {
  override val modifierDelegate: Modifiable
    get() = parent as FieldDeclarationGroup
}
```
Pros: concise; `modifiers` is `val` so it can't be modified indirectly
Cons: `DelegatesModifiers` has to be modified when the interface of `Modifiable` changes

---

3. Mixin-style interface

e.g.
```kotlin
interface DelegatesModifiers : Modifiable {
  val modifierDelegate: Modifiable

  override var modifiers: List<Modifier>
    get() = modifierDelegate.modifiers
    set(value) {
      modifierDelegate.modifiers = value
    }
}
```
where `FieldDeclaration` is the same as above.

Pros: concise; interface changes are automatically reflected
Cons: `modifiers` is `var`

---

4. Implement `Modifiable` directly

e.g.
```kotlin
class FieldDeclaration : Declaration(), Modifiable {
  private val modifierDelegate: Modifiable // kept property for convenience but not necessary
    get() = parent as FieldDeclarationGroup

  override var modifiers: List<Modifier>
    get() = modifierDelegate.modifiers
    set(value) {
      modifierDelegate.modifiers = value
    }
}
```
Pros: simplest code
Cons: duplication of delegation logic; `modifiers` is `var`

Reply:
1 is definitely concise. I wouldn't over-optimize for future changes to the grammar: Apex is a pretty stable language.

2 and 3 are workable, but the level of indirection is a readability negative (that I think can be avoided).

4 also looks nice and simple to me. For this solution, is there any way to make `set` private (or otherwise make modifiers effectively immutable)?

Reply:
1: I mainly had in mind variable declarations, which should mirror the `FieldDeclarationGroup`-`FieldDeclaration` structure. I'm not sure whether it would be possible to refactor `Translate` to support this design for variables.

2 and 3: agree

4: It is not possible to set the setter visibility lower than the visibility in `Modifiable`, which has to be `public` for `Translate` to be able to initialize modifiers. It is also not possible to override the `var modifiers` with `val`. (If `Translate` was able to pass modifiers to each node constructor immediately, `Modifiable.modifiers` could be `val` and this would be a lot simpler.) 

To make `modifiers` effectively immutable, the setter could be empty:
```kotlin
override var modifiers: List<Modifier>
  get() = modifierDelegate.modifiers
  set(_) {}
```
I was already aware of this option, but it doesn't seem clean to indicate to clients that `modifiers` is mutable while the setter performs no operation.

Reply:
1. I am sure we can figure this out. A pattern I used before was to return transient data classes to help accumulate all of the information to build e.g. a `VariableDeclarationStatement` and then actually construct that object once all of the data is available in the transient data class.

4. Yup, I agree empty setter is unfriendly. An exception is slightly better, but not ideal. Maybe @zorzella has an opinion?

Reply:
I'll go with my 2cs here, rather than an opinion on which choice to take:

wrt 4's listed cons: "duplication of delegation logic; modifiers is var". The duplication looks trivial, and does not really bother me. The "var" con is a bit more bothersome

wrt 2's listed con: "DelegatesModifiers has to be modified when the interface of Modifiable changes", I'm not quite sure I would even call that a con. As for Aaron's point "the level of indirection is a readability negative", at a first glance that does not seem to me to be such a problem, though I might be missing something wrt to how this complexity multiplies with other intrinsic complexity. And the "val" seems to be a compelling reason to take this choice.

I defer to your judgement in this regard, though; none of the options seem egregiously bad...

Reply:
Thanks for the opinions. My new thought is that `FieldDeclaration` seeing `modifiers` as `var` isn't actually directly an issue, but a symptom of `modifiers` being mutable in `Modifiable`. In the future, I think this aspect should be changed (`Translate` will have to be refactored), but for now I will disregard this as a factor.

In that case, option 2 can be eliminated, because its only benefit over 3 is that `modifiers` is a `val`.

I dislike option 1 because it imposes a certain translation order that happens to be the opposite of how all other nodes are translated. I would rather not introduce coupling between these two parts of the program, especially when there are other solutions that are relatively good.

Between 3 and 4, I agree that the duplication is trivial and maybe it's not worth creating an additional interface that increases the complexity of the code. Regardless, they are very similar solutions. If/when `modifiers` is changed to `val`, the implementation for both cases would be even simpler:
```kotlin
override val modifiers: List<Modifier>
  get() = modifierDelegate.modifiers
```
For now, I will go with option 4 - implementing `Modifiable` directly.